File: ios/RoutineChart/RoutineChart/Features/FamilyInvite/GenerateInviteViewModel.swift
//
//  GenerateInviteViewModel.swift
//  RoutineChart
//
//  Created for Phase 2.2: QR Family Joining
//

import Foundation
import UIKit
import Combine
import OSLog

@MainActor
final class GenerateInviteViewModel: ObservableObject {
    @Published var qrCodeImage: UIImage?
    @Published var invite: FamilyInvite?
    @Published var errorMessage: String?
    @Published var isLoading = false
    @Published var loadingMessage = "Loading..."
    @Published var timeRemaining: String = ""
    
    private let inviteRepository: FamilyInviteRepository
    private let familyRepository: FamilyRepository
    private let userRepository: UserRepository
    private let authRepository: AuthRepository
    private var timer: Timer?
    private var cancellables = Set<AnyCancellable>()
    private var inviteListener: FirestoreInviteListener?
    
    init(
        inviteRepository: FamilyInviteRepository,
        familyRepository: FamilyRepository,
        userRepository: UserRepository,
        authRepository: AuthRepository
    ) {
        self.inviteRepository = inviteRepository
        self.familyRepository = familyRepository
        self.userRepository = userRepository
        self.authRepository = authRepository
    }
    
    deinit {
        timer?.invalidate()
        // Note: inviteListener will clean itself up via its own deinit
        // We can't call stopRealTimeListener() here because deinit is not @MainActor isolated
    }
    
    func loadActiveInvite() async {
        isLoading = true
        loadingMessage = "Loading..."
        errorMessage = nil
        
        do {
            // Get the current user's familyId (source of truth)
            guard let authUser = authRepository.currentUser else {
                isLoading = false
                return
            }
            
            guard let user = try await userRepository.get(id: authUser.id) else {
                AppLogger.ui.error("No user record found for authenticated user: \(authUser.id)")
                isLoading = false
                return
            }
            
            let familyId = user.familyId
            AppLogger.ui.info("Loading invites for user's family: \(familyId)")
            
            // Sync invites from Firestore first (to get invites created on other devices)
            if let compositeRepo = inviteRepository as? CompositeFamilyInviteRepository {
                do {
                    try await compositeRepo.syncFromFirestore(familyId: familyId)
                    AppLogger.ui.info("Synced invites from Firestore for family: \(familyId)")
                } catch {
                    // Log but don't fail - we can still use local data
                    AppLogger.ui.error("Failed to sync invites from Firestore: \(error.localizedDescription)")
                }
            }
            
            // Get all active invites for this family (now includes synced invites)
            let activeInvites = try await inviteRepository.getActiveInvites(familyId: familyId)
            
            // Find the first valid (not expired, not max uses) invite
            if let validInvite = activeInvites.first(where: { $0.isValid }) {
                self.invite = validInvite
                self.qrCodeImage = QRCodeGenerator.generate(for: validInvite)
                startTimer()
                startRealTimeListener(for: validInvite)
                AppLogger.ui.info("Loaded existing active invite: \(validInvite.id)")
            }
        } catch {
            AppLogger.ui.error("Failed to load active invite: \(error.localizedDescription)")
            // Don't show error to user - just proceed to generate new one if needed
        }
        
        isLoading = false
    }
    
    func generateInvite() async {
        isLoading = true
        loadingMessage = "Generating invite..."
        errorMessage = nil
        
        do {
            // Get the current user's familyId (source of truth)
            guard let authUser = authRepository.currentUser else {
                errorMessage = "Please sign in to generate an invite"
                isLoading = false
                return
            }
            
            guard let user = try await userRepository.get(id: authUser.id) else {
                errorMessage = "No user record found. Please sign in again."
                isLoading = false
                return
            }
            
            let familyId = user.familyId
            AppLogger.ui.info("Generating invite for user's family: \(familyId)")
            
            let createdBy = authUser.id
            
            // Create invite
            let token = TokenGenerator.generateSecureToken()
            let inviteCode = InviteCodeGenerator.generateInviteCode()
            let expiresAt = Calendar.current.date(byAdding: .hour, value: 24, to: Date()) ?? Date().addingTimeInterval(86400)
            
            let newInvite = FamilyInvite(
                familyId: familyId,
                token: token,
                inviteCode: inviteCode,
                createdBy: createdBy,
                expiresAt: expiresAt
            )
            
            try await inviteRepository.create(newInvite)
            self.invite = newInvite
            
            // Generate QR code
            self.qrCodeImage = QRCodeGenerator.generate(for: newInvite)
            
            // Start timer to update time remaining
            startTimer()
            
            // Start real-time listener for updates
            startRealTimeListener(for: newInvite)
            
            AppLogger.ui.info("Generated family invite: \(newInvite.id)")
        } catch {
            errorMessage = "Failed to generate invite: \(error.localizedDescription)"
            AppLogger.ui.error("Failed to generate invite: \(error)")
        }
        
        isLoading = false
    }
    
    func deactivateInvite() async {
        guard let invite = invite else { return }
        
        do {
            try await inviteRepository.deactivate(id: invite.id)
            stopRealTimeListener()
            self.invite = nil
            self.qrCodeImage = nil
            timer?.invalidate()
            AppLogger.ui.info("Deactivated invite: \(invite.id)")
        } catch {
            errorMessage = "Failed to deactivate invite: \(error.localizedDescription)"
            AppLogger.ui.error("Failed to deactivate invite: \(error)")
        }
    }
    
    func shareableImage(for invite: FamilyInvite, qrImage: UIImage) -> UIImage? {
        // Create a composite image with QR code and invite code text
        let size = CGSize(width: 600, height: 800)
        let renderer = UIGraphicsImageRenderer(size: size)
        
        return renderer.image { context in
            // White background
            UIColor.white.setFill()
            context.fill(CGRect(origin: .zero, size: size))
            
            // Title text
            let titleText = "Join my family on Routine Chart!"
            let titleAttributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 28, weight: .bold),
                .foregroundColor: UIColor.label
            ]
            let titleSize = titleText.size(withAttributes: titleAttributes)
            let titleRect = CGRect(
                x: (size.width - titleSize.width) / 2,
                y: 40,
                width: titleSize.width,
                height: titleSize.height
            )
            titleText.draw(in: titleRect, withAttributes: titleAttributes)
            
            // Invite code text
            let codeText = "Invite Code: \(invite.inviteCode)"
            let codeAttributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.monospacedSystemFont(ofSize: 32, weight: .bold),
                .foregroundColor: UIColor.systemBlue
            ]
            let codeSize = codeText.size(withAttributes: codeAttributes)
            let codeRect = CGRect(
                x: (size.width - codeSize.width) / 2,
                y: titleRect.maxY + 30,
                width: codeSize.width,
                height: codeSize.height
            )
            codeText.draw(in: codeRect, withAttributes: codeAttributes)
            
            // QR Code (centered, larger)
            let qrSize: CGFloat = 400
            let qrRect = CGRect(
                x: (size.width - qrSize) / 2,
                y: codeRect.maxY + 40,
                width: qrSize,
                height: qrSize
            )
            qrImage.draw(in: qrRect)
            
            // Instructions text
            let instructionText = "Scan the QR code or use the invite code above"
            let instructionAttributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 18, weight: .regular),
                .foregroundColor: UIColor.secondaryLabel
            ]
            let instructionSize = instructionText.size(withAttributes: instructionAttributes)
            let instructionRect = CGRect(
                x: (size.width - instructionSize.width) / 2,
                y: qrRect.maxY + 30,
                width: instructionSize.width,
                height: instructionSize.height
            )
            instructionText.draw(in: instructionRect, withAttributes: instructionAttributes)
        }
    }
    
    private func startTimer() {
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                self?.updateTimeRemaining()
            }
        }
        updateTimeRemaining()
    }
    
    private func updateTimeRemaining() {
        guard let invite = invite else {
            timeRemaining = ""
            return
        }
        
        let remaining = invite.timeRemaining
        if remaining <= 0 {
            timeRemaining = "Expired"
            timer?.invalidate()
            // Auto-dismiss expired invites
            stopRealTimeListener()
            self.invite = nil
            self.qrCodeImage = nil
            return
        }
        
        let hours = Int(remaining) / 3600
        let minutes = (Int(remaining) % 3600) / 60
        let seconds = Int(remaining) % 60
        
        if hours > 0 {
            timeRemaining = String(format: "Expires in %dh %dm", hours, minutes)
        } else if minutes > 0 {
            timeRemaining = String(format: "Expires in %dm %ds", minutes, seconds)
        } else {
            timeRemaining = String(format: "Expires in %ds", seconds)
        }
    }
    
    // MARK: - Real-time Updates
    
    private func startRealTimeListener(for invite: FamilyInvite) {
        // Stop any existing listener
        stopRealTimeListener()
        
        // Create new listener
        let listener = FirestoreInviteListener()
        self.inviteListener = listener
        
        // Subscribe to updates
        listener.invitePublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] updatedInvite in
                guard let self = self else { return }
                
                if let updatedInvite = updatedInvite {
                    // Update the invite with real-time data
                    self.invite = updatedInvite
                    
                    // If invite is no longer valid (expired, deactivated, max uses), clear it
                    if !updatedInvite.isValid {
                        AppLogger.ui.info("Invite \(updatedInvite.id) is no longer valid - clearing")
                        self.stopRealTimeListener()
                        self.invite = nil
                        self.qrCodeImage = nil
                        self.timer?.invalidate()
                    }
                } else {
                    // Invite was deleted or doesn't exist
                    AppLogger.ui.info("Invite was deleted or doesn't exist - clearing")
                    self.stopRealTimeListener()
                    self.invite = nil
                    self.qrCodeImage = nil
                    self.timer?.invalidate()
                }
            }
            .store(in: &cancellables)
        
        // Start listening
        listener.startListening(inviteId: invite.id, familyId: invite.familyId)
        AppLogger.ui.info("Started real-time listener for invite: \(invite.id)")
    }
    
    private func stopRealTimeListener() {
        inviteListener?.stopListening()
        inviteListener = nil
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/features/familyinvite/GenerateInviteViewModel.kt
package com.HammersTech.RoutineChart.features.familyinvite

import android.graphics.Bitmap
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.HammersTech.RoutineChart.core.domain.models.FamilyInvite
import com.HammersTech.RoutineChart.core.domain.repositories.FamilyInviteRepository
import com.HammersTech.RoutineChart.core.domain.repositories.FamilyRepository
import com.HammersTech.RoutineChart.core.utils.AppLogger
import com.HammersTech.RoutineChart.core.utils.InviteCodeGenerator
import com.HammersTech.RoutineChart.core.utils.QRCodeGenerator
import com.HammersTech.RoutineChart.core.utils.TokenGenerator
import com.HammersTech.RoutineChart.core.utils.ULIDGenerator
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import javax.inject.Inject

/**
 * ViewModel for generating and displaying family invites
 * Phase 2.2: QR Family Joining
 */
@HiltViewModel
class GenerateInviteViewModel
    @Inject
    constructor(
        private val inviteRepository: FamilyInviteRepository,
        private val familyRepository: FamilyRepository,
        private val userRepository: com.HammersTech.RoutineChart.core.domain.repositories.UserRepository,
        private val authRepository: com.HammersTech.RoutineChart.core.domain.repositories.AuthRepository,
    ) : ViewModel() {
        data class UiState(
            val isLoading: Boolean = false,
            val loadingMessage: String = "Loading...",
            val qrCodeBitmap: Bitmap? = null,
            val invite: FamilyInvite? = null,
            val errorMessage: String? = null,
            val timeRemaining: String = "",
        )

        private val _state = MutableStateFlow(UiState())
        val state: StateFlow<UiState> = _state.asStateFlow()

        private var timerJob: kotlinx.coroutines.Job? = null

        fun loadActiveInvite() {
            viewModelScope.launch {
                _state.value =
                    _state.value.copy(
                        isLoading = true,
                        loadingMessage = "Loading...",
                        errorMessage = null,
                    )

                try {
                    // Get current authenticated user
                    val authUser =
                        authRepository.currentUser
                            ?: throw IllegalStateException("Not authenticated")

                    // Get current user's familyId (source of truth)
                    val currentUser = userRepository.getById(authUser.id)
                    if (currentUser == null) {
                        _state.value = _state.value.copy(isLoading = false)
                        return@launch
                    }

                    val currentFamilyId = currentUser.familyId

                    // Get the current family
                    val family = familyRepository.getById(currentFamilyId)
                    if (family == null) {
                        _state.value = _state.value.copy(isLoading = false)
                        return@launch
                    }

                    // Sync invites from Firestore first (to get invites created on other devices)
                    if (inviteRepository is com.HammersTech.RoutineChart.core.data.remote.firebase.CompositeFamilyInviteRepository) {
                        try {
                            (inviteRepository as com.HammersTech.RoutineChart.core.data.remote.firebase.CompositeFamilyInviteRepository)
                                .syncFromFirestore(currentFamilyId)
                            AppLogger.UI.info("Synced invites from Firestore for family: $currentFamilyId")
                        } catch (e: Exception) {
                            // Log but don't fail - we can still use local data
                            AppLogger.UI.error("Failed to sync invites from Firestore: ${e.message}", e)
                        }
                    }

                    // Get all active invites for this family (now includes synced invites)
                    val activeInvites = inviteRepository.getActiveInvites(currentFamilyId)

                    // Find the first valid (not expired, not max uses) invite
                    val validInvite = activeInvites.firstOrNull { it.isValid }

                    if (validInvite != null) {
                        // Generate QR code
                        val qrCode = QRCodeGenerator.generate(validInvite)

                        _state.value =
                            _state.value.copy(
                                isLoading = false,
                                invite = validInvite,
                                qrCodeBitmap = qrCode,
                            )

                        // Start timer to update time remaining
                        startTimer()
                        AppLogger.UI.info("Loaded existing active invite: ${validInvite.id}")
                    }
                } catch (e: Exception) {
                    AppLogger.UI.error("Failed to load active invite", e)
                    // Don't show error to user - just proceed to generate new one if needed
                }

                _state.value = _state.value.copy(isLoading = false)
            }
        }

        fun generateInvite() {
            viewModelScope.launch {
                _state.value =
                    _state.value.copy(
                        isLoading = true,
                        loadingMessage = "Generating invite...",
                        errorMessage = null,
                    )

                try {
                    // Get current authenticated user
                    val authUser =
                        authRepository.currentUser
                            ?: throw IllegalStateException("Not authenticated")

                    // Get current user's familyId (source of truth)
                    val currentUser = userRepository.getById(authUser.id)
                    if (currentUser == null) {
                        _state.value =
                            _state.value.copy(
                                isLoading = false,
                                errorMessage = "User not found. Please sign in again.",
                            )
                        return@launch
                    }

                    val currentFamilyId = currentUser.familyId

                    // Get the current family
                    val family = familyRepository.getById(currentFamilyId)
                    if (family == null) {
                        _state.value =
                            _state.value.copy(
                                isLoading = false,
                                errorMessage = "No family found. Please create a family first.",
                            )
                        return@launch
                    }

                    // Use the authUser we already got above
                    val createdBy = authUser.id

                    // Create invite
                    val token = TokenGenerator.generateSecureToken()
                    val inviteCode = InviteCodeGenerator.generateInviteCode()
                    val expiresAt = Instant.now().plusSeconds(86400) // 24 hours

                    val newInvite =
                        FamilyInvite(
                            id = ULIDGenerator.generate(),
                            familyId = family.id,
                            token = token,
                            inviteCode = inviteCode,
                            createdBy = createdBy,
                            expiresAt = expiresAt,
                        )

                    inviteRepository.create(newInvite)

                    // Generate QR code
                    val qrCode = QRCodeGenerator.generate(newInvite)

                    _state.value =
                        _state.value.copy(
                            isLoading = false,
                            invite = newInvite,
                            qrCodeBitmap = qrCode,
                        )

                    // Start timer to update time remaining
                    startTimer()

                    AppLogger.UI.info("Generated family invite: ${newInvite.id}")
                } catch (e: Exception) {
                    AppLogger.UI.error("Failed to generate invite", e)
                    _state.value =
                        _state.value.copy(
                            isLoading = false,
                            errorMessage = "Failed to generate invite: ${e.message}",
                        )
                }
            }
        }

        fun deactivateInvite() {
            viewModelScope.launch {
                val invite = _state.value.invite ?: return@launch

                try {
                    inviteRepository.deactivate(invite.id)
                    _state.value = UiState() // Reset state
                    timerJob?.cancel()
                    AppLogger.UI.info("Deactivated invite: ${invite.id}")
                } catch (e: Exception) {
                    AppLogger.UI.error("Failed to deactivate invite", e)
                    _state.value =
                        _state.value.copy(
                            errorMessage = "Failed to deactivate invite: ${e.message}",
                        )
                }
            }
        }

        fun shareableImage(
            invite: FamilyInvite,
            qrBitmap: Bitmap,
        ): Bitmap? {
            return try {
                // Create a composite image with QR code and invite code text
                val width = 600
                val height = 800
                val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
                val canvas = android.graphics.Canvas(bitmap)

                // White background
                canvas.drawColor(android.graphics.Color.WHITE)

                val paint =
                    android.graphics.Paint().apply {
                        isAntiAlias = true
                        textAlign = android.graphics.Paint.Align.CENTER
                    }

                var yPos = 60f

                // Title text
                paint.textSize = 28f
                paint.typeface = android.graphics.Typeface.create(android.graphics.Typeface.DEFAULT, android.graphics.Typeface.BOLD)
                paint.color = android.graphics.Color.BLACK
                canvas.drawText("Join my family on Routine Chart!", width / 2f, yPos, paint)
                yPos += 50f

                // Invite code text
                paint.textSize = 32f
                paint.typeface = android.graphics.Typeface.create(android.graphics.Typeface.MONOSPACE, android.graphics.Typeface.BOLD)
                paint.color = android.graphics.Color.parseColor("#2196F3") // Material Blue
                canvas.drawText("Invite Code: ${invite.inviteCode}", width / 2f, yPos, paint)
                yPos += 60f

                // QR Code (centered, larger)
                val qrSize = 400
                val qrX = (width - qrSize) / 2f
                val qrY = yPos
                canvas.drawBitmap(
                    Bitmap.createScaledBitmap(qrBitmap, qrSize, qrSize, true),
                    qrX,
                    qrY,
                    null,
                )
                yPos += qrSize + 40f

                // Instructions text
                paint.textSize = 18f
                paint.typeface = android.graphics.Typeface.create(android.graphics.Typeface.DEFAULT, android.graphics.Typeface.NORMAL)
                paint.color = android.graphics.Color.GRAY
                canvas.drawText("Scan the QR code or use the invite code above", width / 2f, yPos, paint)

                bitmap
            } catch (e: Exception) {
                AppLogger.UI.error("Failed to create shareable image", e)
                null
            }
        }

        private fun startTimer() {
            timerJob?.cancel()
            timerJob =
                viewModelScope.launch {
                    while (true) {
                        updateTimeRemaining()
                        delay(1000) // Update every second
                    }
                }
        }

        private fun updateTimeRemaining() {
            val invite = _state.value.invite ?: return

            val remaining = invite.timeRemaining
            if (remaining <= 0) {
                _state.value = _state.value.copy(timeRemaining = "Expired")
                timerJob?.cancel()
                return
            }

            val hours = remaining / 3600
            val minutes = (remaining % 3600) / 60
            val seconds = remaining % 60

            val timeString =
                when {
                    hours > 0 -> "Expires in ${hours}h ${minutes}m"
                    minutes > 0 -> "Expires in ${minutes}m ${seconds}s"
                    else -> "Expires in ${seconds}s"
                }

            _state.value = _state.value.copy(timeRemaining = timeString)
        }

        override fun onCleared() {
            super.onCleared()
            timerJob?.cancel()
        }
    }
